### PR TITLE
Fix network memory leaks

### PR DESCRIPTION
**Pull request by asonnino**
Fix memory leaks  in the network stack when updating the committee (or changing epoch)

### CODE REVIEW

File: network/src/primary.rs [hunk 1]
```diff
@@ -73,6 +73,15 @@ impl PrimaryNetwork {
...
+
+    pub fn cleanup<'a, I>(&mut self, to_remove: I)
+    where
+        I: IntoIterator<Item = &'a Multiaddr>,
+    {
+        for address in to_remove {
+            self.clients.remove(address);
+        }
+    }
 }
```
Comment (akichidis) [inline network/src/primary.rs:85]: Could you please add tests for this behaviour?

**Review by akichidis — COMMENTED** [review]
I would like to have a look again on this, but don't have many comments. Tried to think of how we could probably implement in a different way (seems to me that we do similar stuff in many places) but can't think of something right now. In the meanwhile could you please sync with @arun-koshy as he is looking to extra the worker addresses from committee? It seems that you touch some similar stuff.

File: network/src/primary.rs (hunk 1 above)
Comment (asonnino) [reply network/src/primary.rs:85]: @huitseeker Do you think this function could be in a nice trait? It is used by 3 out of 4 of our network structs

**Review by bmwill — APPROVED** [review]

File: network/src/primary.rs (hunk 1 above)
Comment (huitseeker) [reply network/src/primary.rs:85]: @asonnino The natural place to add it is the `BaseNetwork` trait.

File: config/src/lib.rs [hunk 2]
```diff
@@ -451,6 +451,45 @@ impl Committee {
...
+                            .map(|address| &address.primary_to_worker),
+                    )
+            })
+            .flatten()
+            .collect()
+    }
+
+    /// Return the network addresses that are present in the current committee but that are absent
+    /// from the new committee (provided as argument).
+    pub fn network_diff<'a>(&'a self, new_committee: &'a Self) -> HashSet<&Multiaddr> {
```
Comment (huitseeker) [inline config/src/lib.rs:486]: Nit: `new` and `old` are concepts for the caller of this method. The typical way to present things in such a method is `self` and `other` (for the argument).

**Review by huitseeker — COMMENTED** [review]
Same comment on `get_all_network_addresses`, otherwise LGTM. Please link to #566, which you're closing.

**Comment by huitseeker** [comment]
@asonnino You'll want to rebase as we have fixed your `cargo-deny` issue on main.

File: network/src/primary.rs (hunk 1 above)
Comment (asonnino) [reply network/src/primary.rs:85]: But not all networks can have this functions. To be precise, they all need it except `WorkerToPrimary` that only has a single client.

**Pull request merged**